### DISCUSSION TITLE
Independent Authentication Guard - log in issue

### DISCUSSION THREAD

**Discussion by web541** [statamic/cms#10450] [Troubleshooting] [answered]
Hi, I'm testing out to see whether it's possible to add Statamic to an existing Laravel app, but have the Laravel users separate (in a DB) to the Statamic users (in files) as we would only have one or a few users actually editing the Statamic side through the /cp.

I've been following this guide exactly: [https://statamic.dev/tips/using-an-independent-authentication-guard](https://statamic.dev/tips/using-an-independent-authentication-guard) but I seem to be getting this error:
```
call_user_func_array(): Argument #1 ($callback) must be a valid callback, first array member is not a valid class name or object
```
<img width="1201" alt="demo" src="https://github.com/user-attachments/assets/649413d4-624e-4352-9708-8ce3bc614b0f">

After some further digging, it looks like it was still trying to use the eloquent driver and was expecting a 'model' in the config/auth.php `providers` array for the statamic provider which didn't seem right to me:
```php
'providers' => [
    'users' => [
        'driver' => 'eloquent',
        'model'  => \App\Models\User::class,
    ],

    'statamic' => [
        'driver' => 'statamic',
       // expecting a 'model' here...
    ],
],
```

However, I figured out that it was still using the eloquent repository in `config/statamic/users.php`, so swapping this over to `file` does fix the error and the /cp login form appears to work except that it doesn't actually log the user in (despite the user existing in the /users directory) and acts as if the credentials could not be found. This part wasn't mentioned in the docs, so I wasn't sure if this was a necessary step, but figured it was worth a shot.
```php
'repository' => 'file',
```

All of this was tested on a brand new Laravel install (v11.15.0) with Statamic installed using this guide [https://statamic.dev/installing/laravel](https://statamic.dev/installing/laravel) before I try it on a more developed project.

Any thoughts? Thanks!

**Comment by huuvan20** [comment]
The same issue for me which is the login page accepts the credentials (which is the correct one) but nothing else happens after that. The login form is empty without letting users in.
@web541 Have you been able to fix that yet?

**Reply by web541** [reply]
Not yet, I was hoping someone from the team might be able to chime in and provide some guidance. I only got as far as what was in the post above, not sure where to look next.

**Reply by duncanmcclean** [reply]
Sorry, I missed this issue the first time around.

Are either of you able to share the contents of your `config/auth.php` and `config/statamic/users.php` config files? Also, did you install Statamic fresh or into an existing Laravel app?

**Reply by huuvan20** [reply]
Hi @duncanmcclean, sure things

config/auth.php

```
<?php

return [

    /*
    |--------------------------------------------------------------------------
    | Authentication Defaults
    |--------------------------------------------------------------------------
    |
    | This option defines the default authentication "guard" and password
    | reset "broker" for your application. You may change these values
    | as required, but they're a perfect start for most applications.
    |
    */

    'defaults' => [
        'guard' => env('AUTH_GUARD', 'web'),
        'passwords' => env('AUTH_PASSWORD_BROKER', 'users'),
    ],

    /*
    |--------------------------------------------------------------------------
    | Authentication Guards
    |--------------------------------------------------------------------------
    |
    | Next, you may define every authentication guard for your application.
    | Of course, a great default configuration has been defined for you
    | which utilizes session storage plus the Eloquent user provider.
    |
    | All authentication guards have a user provider, which defines how the
    | users are actually retrieved out of your database or other storage
    | system used by the application. Typically, Eloquent is utilized.
    |
    | Supported: "session"
    |
    */

    'guards' => [
        'web' => [
            'driver' => 'session',
            'provider' => 'users',
        ],

        'statamic' => [
            'driver' => 'session',
            'provider' => 'statamic',
        ]
    ],

    /*
    |--------------------------------------------------------------------------
    | User Providers
    |--------------------------------------------------------------------------
    |
    | All authentication guards have a user provider, which defines how the
    | users are actually retrieved out of your database or other storage
    | system used by the application. Typically, Eloquent is utilized.
    |
    | If you have multiple user tables or models you may configure multiple
    | providers to represent the model / table. These providers may then
    | be assigned to any extra authentication guards you have defined.
    |
    | Supported: "database", "eloquent"
    |
    */

    'providers' => [
        'users' => [
            'driver' => 'eloquent',
            'model' => env('AUTH_MODEL', App\Models\User::class),
        ],

        'statamic' => [
            'driver' => 'statamic',
        ],
    ],

    /*
    |--------------------------------------------------------------------------
    | Resetting Passwords
    |--------------------------------------------------------------------------
    |
    | These configuration options specify the behavior of Laravel's password
    | reset functionality, including the table utilized for token storage
    | and the user provider that is invoked to actually retrieve users.
    |
    | The expiry time is the number of minutes that each reset token will be
    | considered valid. This security feature keeps tokens short-lived so
    | they have less time to be guessed. You may change this as needed.
    |
    | The throttle setting is the number of seconds a user must wait before
    | generating more password reset tokens. This prevents the user from
    | quickly generating a very large amount of password reset tokens.
    |
    */

    'passwords' => [
        'users' => [
            'provider' => 'users',
            'table' => env('AUTH_PASSWORD_RESET_TOKEN_TABLE', 'password_reset_tokens'),
            'expire' => 60,
            'throttle' => 60,
        ],

        'statamic_resets' => [
            'provider' => 'statamic',
            'table' => 'password_resets',
            'expire' => 60,
            'throttle' => 60,
        ],

        'statamic_activations' => [
            'provider' => 'statamic',
            'table' => 'password_activations',
            'expire' => 4320,
            'throttle' => 60,
        ],
    ],

    /*
    |--------------------------------------------------------------------------
    | Password Confirmation Timeout
    |--------------------------------------------------------------------------
    |
    | Here you may define the amount of seconds before a password confirmation
    | window expires and users are asked to re-enter their password via the
    | confirmation screen. By default, the timeout lasts for three hours.
    |
    */

    'password_timeout' => env('AUTH_PASSWORD_TIMEOUT', 10800),

];

```
config/statamic/users.php
```
<?php

return [

    /*
    |--------------------------------------------------------------------------
    | User Repository
    |--------------------------------------------------------------------------
    |
    | Statamic uses a repository to get users, roles, groups, and their
    | relationships from specified storage locations. The file driver
    | gets it from disk, while the eloquent driver gets from a DB.
    |
    | Supported: "file", "eloquent"
    |
    */

    'repository' => 'file',

    'repositories' => [

        'file' => [
            'driver' => 'file',
            'paths' => [
                'roles' => resource_path('users/roles.yaml'),
                'groups' => resource_path('users/groups.yaml'),
            ],
        ],

        'eloquent' => [
            'driver' => 'eloquent',
        ],

    ],

    /*
    |--------------------------------------------------------------------------
    | Avatars
    |--------------------------------------------------------------------------
    |
    | User avatars are initials by default, with custom options for services
    | like Gravatar.com.
    |
    | Supported: "initials", "gravatar", or a custom class name.
    |
    */

    'avatars' => 'initials',

    /*
    |--------------------------------------------------------------------------
    | New User Roles
    |--------------------------------------------------------------------------
    |
    | When registering new users through the user:register_form tag, these
    | roles will automatically be applied to your newly created users.
    |
    */

    'new_user_roles' => [
        //
    ],

    /*
    |--------------------------------------------------------------------------
    | New User Groups
    |--------------------------------------------------------------------------
    |
    | When registering new users through the user:register_form tag, these
    | groups will automatically be applied to your newly created users.
    |
    */

    'new_user_groups' => [
        //
    ],

    /*
    |--------------------------------------------------------------------------
    | Registration form honeypot field
    |--------------------------------------------------------------------------
    |
    | When registering new users through the user:register_form tag,
    | specify the field to act as a honeypot for bots
    |
    */

    'registration_form_honeypot_field' => null,

    /*
    |--------------------------------------------------------------------------
    | User Wizard Invitation Email
    |--------------------------------------------------------------------------
    |
    | When creating new users through the wizard in the control panel,
    | you may choose whether to be able to send an invitation email.
    | Setting to true will give the user the option. But setting
    | it to false will disable the invitation option entirely.
    |
    */

    'wizard_invitation' => true,

    /*
    |--------------------------------------------------------------------------
    | Password Brokers
    |--------------------------------------------------------------------------
    |
    | When resetting passwords, Statamic uses an appropriate password broker.
    | Here you may define which broker should be used for each situation.
    | You may want a longer expiry for user activations, for example.
    |
    */

    'passwords' => [
        'resets' => 'statamic_resets',
        'activations' => 'statamic_activations',
    ],

    /*
    |--------------------------------------------------------------------------
    | Database
    |--------------------------------------------------------------------------
    |
    | Here you may configure the database connection and its table names.
    |
    */

    'database' => config('database.default'),

    'tables' => [
        'users' => 'users',
        'role_user' => 'role_user',
        'roles' => false,
        'group_user' => 'group_user',
        'groups' => false,
    ],

    /*
    |--------------------------------------------------------------------------
    | Authentication Guards
    |--------------------------------------------------------------------------
    |
    | By default, Statamic will use the `web` authentication guard. However,
    | if you want to run Statamic alongside the default Laravel auth
    | guard, you can configure that for your cp and/or frontend.
    |
    */

    'guards' => [
        'cp' => 'statamic',
        'web' => 'web',
    ],

    /*
    |--------------------------------------------------------------------------
    | Impersonation
    |--------------------------------------------------------------------------
    |
    | Here you can configure if impersonation is available, and what URL to
    | redirect to after impersonation begins.
    |
    */

    'impersonate' => [
        'enabled' => env('STATAMIC_IMPERSONATE_ENABLED', true),
        'redirect' => env('STATAMIC_IMPERSONATE_REDIRECT', null),
    ],

    /*
    |--------------------------------------------------------------------------
    | Default Sorting
    |--------------------------------------------------------------------------
    |
    | Here you may configure the default sort behavior for user listings.
    |
    */

    'sort_field' => 'email',
    'sort_direction' => 'asc',

];

```

**Reply by web541** [reply]
Thanks @duncanmcclean Here are mine, similar to the above except I've got the two separate reset and activation brokers as per the dev guide and still have the issue. [config/auth.php](https://gist.github.com/web541/97c746f9b6aa65e41ede66c50a9edd70) and [config/statamic/users.php](https://gist.github.com/web541/c83e5342ee33da2c5c50fc92986bedaa)

I wanted to emulate installing Statamic into an existing Laravel app, so I installed a fresh version of Laravel and then Statamic on top of it as per [this guide](https://statamic.dev/installing/laravel) (so not using the CLI) + Breeze/Jetstream (I tried both). But everything was installed completely fresh, so technically no existing app yet.

**Reply by duncanmcclean** [reply]
Thanks for sharing your config files.

Strangely, even after copying & pasting the contents of both files into my fresh Jetstream app, I can login to the Control Panel without any issues. 🤔

A couple of questions:

1. After configuring Statamic to use flat-file users, did you create a user with `php please make:user`?
2. Are you seeing this error when you try to login to Statamic, or your app? 
3. If you're seeing it when you login to the CP, are you also logged into your app at the same time?

**Reply by web541** [reply]
Hmm, interesting. I just did a full fresh install of Jetstream, then Statamic and I still seem to not be able to log in. Here is a [demo app](https://github.com/web541/statamic-auth-demo) (which I will delete afterwards), but I'll guess it will probably work for you. It could very well be something I'm doing, I'm not sure. These are the steps I used to create it:
- `composer create-project laravel/laravel statamic-auth-demo`
- `composer require laravel/jetstream`
- `php artisan jetstream:install inertia`
- `npm install`
- `npm run build`
- `php artisan migrate`
- Register a new user in the app
- `php artisan config:clear`
- Add `@php artisan statamic:install --ansi` to composer.json as per docs.
- `composer require statamic/cms --with-dependencies`
- Copy over both config files from my reply above
- `php please make:user` for a new superuser for the CP
- Attempt to login to app (works fine)
- Log out
- Attempt to login to CP (doesn't log user in)

To your questions:
1. Yes. Though I noticed the behaviour is slightly different depending on whether the user has the same email in both the /users and in the DB. If the email is the same, it just does the form 'shake' response and acts like it the given credentials were invalid. If the email is different, logging in through the CP with the correct credentials just redirects to /login and wants me to login to the Jetstream app. Weird, could be a session thing?

2. Error appears when attempting to login to Statamic through the CP. But only when config/statamic/users.php `repository` is `eloquent` and not `file` (which is what I assumed it should be, but that's not written in any of the Statamic docs anywhere so I was just testing this blind). When it's set to `file`, it just acts like the credentials are incorrect (even if they're correct) and doesn't let me in.

3. No, it happens when I log into the CP but I'm not currently logged into either. I just tried logging into the app as a test, then the CP, and it just logs me out of the app.

Hope that helps some more!

**Reply by duncanmcclean** [reply]
> Error appears when attempting to login to Statamic through the CP. But only when config/statamic/users.php repository is eloquent and not file (which is what I assumed it should be, but that's not written in any of the Statamic docs anywhere so I was just testing this blind).

Ohh, it'll need to be `file`, otherwise it'll be throwing stuff off. 

Otherwise, it'll try and use your user from the `users` table in the database, but the guard isn't configured to make that work so stuff won't work. We can update the docs.

> When it's set to file, it just acts like the credentials are incorrect (even if they're correct) and doesn't let me in.

When you try to login, is there a `password_hash` key in your user's YAML file, or just a `password` key? Does a password reset work?

**Reply by web541** [reply]
> Ohh, it'll need to be `file`, otherwise it'll be throwing stuff off.

Yep, makes sense.

> When you try to login, is there a `password_hash` key in your user's YAML file, or just a `password` key?

Yes, there's a `password_hash` key. There has never been a `password` key since I used the `php please make:user` command, unless I manually put a password there to be hashed by Statamic later (which I haven't tried).

> Does a password reset work?

Password reset appears to work, but still can't login. I have noticed a change in behaviour though. Once the password is changed, if I use the correct password to login it just refreshes the form (no shaking). But if I then use the old (incorrect) password, the form shakes and acts like the credentials are incorrect, but I'm not sure if it's actually logging me in and not starting the session/redirecting me or if it's not even getting to that stage. But now I can confirm at least that it appears to accept the correct credentials.

**Reply by duncanmcclean** [reply]
Hmm, that's strange!

I just cloned down [your demo repository](https://github.com/web541/statamic-auth-demo) so I'm testing w/ the same code.

I registered a databased user (which I could login to fine with Jetstream), ran `php please make:user` and created a super user. When I login at `/cp`, it logged me in successfully. 🤔 

I'm out of ideas now... I can't really troubleshoot it further if I can't reproduce it 😅. Maybe it's a browser/caching thing? Can you try a diferent browser?

**Reply by web541** [reply]
That is really weird! I've tried various fresh installs and I can't seem to figure out how to get it working. Out of interest, what is your dev stack running on? I'm on valet, but might try Herd and see if that does anything.

> I registered a databased user (which I could login to fine with Jetstream), ran php please make:user and created a super user. When I login at /cp, it logged me in successfully. 🤔

Did both of your users have the same email / credentials?

> Maybe it's a browser/caching thing? Can you try a diferent browser?

Tried Chrome, Safari, Firefox and they're all having the same issue :(

**Reply by duncanmcclean** [reply]
> That is really weird! I've tried various fresh installs and I can't seem to figure out how to get it working. Out of interest, what is your dev stack running on? I'm on valet, but might try Herd and see if that does anything.

I'm using Laravel Herd (which is basically Valet under the hood).

> Did both of your users have the same email / credentials?

No, they had different emails & passwords.

**Reply by jerwre** [reply]
I'm experiencing the same issue. I'm installing statamic into a fresh laravel installation (laravel.build) following the docs, and I change to use file for users in configs. Create a new user via CLI (user file gets created), and when I try to log in I just get sent back. The form does not shake as if I entered the wrong credentials. 

When I installed a new project with the statamic cli, I can switch between file and database, and I can set it up to use file for cp/statamic and database for users in general. I copied over the auth and users config files to the first project just in case, but it did not help.

**Reply by duncanmcclean** [reply]
I've just tested this again... 

I cloned the repository @web541 made and registered a user in Jetstream and created a user with `php please make:user` and I can login to Jetstream/Statamic without any issues. I haven't made changes to any files, apart from creating a Statamic user.

I'm afraid I'm out of ideas. If I can't reproduce it myself, there's not much I can do to debug. Sorry! 😬

**Reply by web54** [reply]
just here to say : it is not me (@web54 ) but @web541 who is the author of this repository

**Reply by duncanmcclean** [reply]
Oops, sorry! 😬

**Reply by jasonvarga** [reply]
That demo site worked perfectly for me too.
- I could log in to the front-end with the DB user that gets seeded (test@example.com / password)
- I could make a Statamic user with `please make:user` and log into `/cp` with it.
- I am logged into the frontend and cp with different users at the same time.

I'll look into some other things though.

**Comment by jasonvarga** [comment]
As for the `call_user_func_array(): Argument #1 ($callback) must be a valid callback, first array member is not a valid class name or object` error in the original post, I was able to trigger this.

- Install Laravel
- Install Statamic into Laravel
- Configure the separate auth guards.
- *Don't* change the `repository` in `config/statamic/users.php`.
- Create a Statamic user with `php please make:user`

Setting the `repository` to `file` makes the error go away.

In my case though, it worked fine at this point. I wasn't able to get the "doesn't actually log the user in" behavior.

I've update the Laravel install and separate auth guard docs in a couple of places to make some things clearer.

**Comment by web541** [comment]
Whoops, just reposting here..

Thanks for looking into this! Just curious, what `DB_CONNECTION` you've been using @jasonvarga @duncanmcclean ? I've been using `mysql` this entire time and it would appear that's where the issues are coming from (I probably should've made that more clear in the .env.example). I just randomly tried it with `sqlite`, seeded using the default user and everything worked perfectly! Was able to log into both Laravel and Statamic /cp simultaneously with two different users from DB and file repositories.

I looked into this further and it could be related to `mysql` and session management. In the .env, this works:
```
DB_CONNECTION=sqlite
..
SESSION_DRIVER=database
```

And this also works:
```
DB_CONNECTION=mysql
..
SESSION_DRIVER=file
```

But this doesn't work (and this is where I get the "doesn't actually log the user in" behavior):
```
DB_CONNECTION=mysql
..
SESSION_DRIVER=database
```
I had a suspicion this might've been the case as when I attempted to log into the /cp, it logged the Laravel user out (in addition to the form not jiggling for wrong credentials and instead just flashed). Any idea why `mysql` connection and the `database` session driver might not be working together?

I have noticed that in the `sessions` table, the `user_id` becomes `null` which may have something to do with it (not 100% though), but when it works using one of the first two .env combos above it the `user_id` column is either filled with the Laravel `user_id` or Statamic `user_id` (guid from file) depending on who has logged in first.

@huuvan20 were you using `mysql` also? If so, would you be able to try what I did above and see if it fixes it for you?

Side note - either way (when it is working), logging out of the /cp also logs out the Laravel user. Is this intended behavior? I'm guessing so based on the single session.

**Reply by duncanmcclean** [reply]
Interesting! 

I think we were both probably testing with SQLite. When I swap SQLite for MySQL, I see the issue (finally! 🎉). 

Can you try making the `user_id` column on the `sessions` table a `varchar`, instead of an integer and let me know if that fixes the issue for you?

> Side note - either way (when it is working), logging out of the /cp also logs out the Laravel user. Is this intended behavior? I'm guessing so based on the single session.

From what I can tell, while probably not ideal, I don't know if there's much we can do about it.

When you logout, [Fortify](https://github.com/laravel/fortify/blob/6516ff91dd58ee7dfb7d5ebd4ae1e24b7a1e5b3b/src/Http/Controllers/AuthenticatedSessionController.php#L100-L110) and [Statamic](https://github.com/statamic/cms/blob/3956ab1343eef8b20d89c5281160c05c63f279c0/src/Http/Controllers/CP/Auth/LoginController.php#L128-L137) invalidate & regenerate the user's session. Since the user only has one session, they're logged out of both guards.

**Reply by web541** [reply]
> I think we were both probably testing with SQLite. When I swap SQLite for MySQL, I see the issue (finally! 🎉).

Definitely my fault, I should've made that more clear!


> Can you try making the `user_id` column on the `sessions` table a `varchar`, instead of an integer and let me know if that fixes the issue for you?

Yep, this has fixed it for me! It appears to take longer to log into the /cp than Laravel, but I think that's more likely my setup than Statamic.


> From what I can tell, while probably not ideal, I don't know if there's much we can do about it.

Yep fair enough, it's very minor anyway and probably wouldn't cause many issues (if any) day-to-day so can easily put up with it.

**Reply by duncanmcclean** [reply, marked as answer]
Great! We can likely just mention in the docs that you may need to change the column type. I've added it to our `docs` repository: https://github.com/statamic/docs/issues/1475

**Reply by AtmoFX** [reply]
@duncanmcclean 
If i am not wrong, the `/database/migrations/` folder contains a file `0001_01_01_000000_create_users_table.php` that is installed when creating a new site. Inside that file, I can see `$table->Id();` for table `users`.

Would it not prevent this issue from arising, had it been `$table->Uuid('id')->primary();` instead?

Changing it removes the `default` value on the `id` column, so something must be added.
The foreign keys on `user_group` and `user_role` would need to be `$table->foreignUuid('user_id')->constrained('users')->cascadeOnDelete();` in the `statamic_auth_tables.php` migration file too.

**Reply by duncanmcclean** [reply]
> Would it not prevent this issue from arising, had it been $table->Uuid('id')->primary(); instead?

I haven't tested it myself, but _potentially_? Although, I don't know that everyone will want to change their database users to use UUIDs to fix the Statamic CP.
